lasio/reader.py

```diff
@@ -751,6 +751,47 @@ def read_header_line(line, pattern=None, section_name=None):
     """
     d = {"name": "", "unit": "", "value": "", "descr": ""}
 
+    # Set defaults for local variables.
+    patterns = []
+    m = None
+
+    if pattern is None:
+        patterns = configure_and_match_pattern(line, section_name)
+    else: # pattern was passed in on function call
+        patterns.append(pattern)
+        m = re.match(pattern, line)
+
+    for pattern in patterns:
+        m = re.match(pattern, line)
+        if m is not None:
+            break
+
+    if m is None:
+        logger.warning("Unable to parse line as LAS header: {}".format(line))
```

Comment:
To clarify something: does this mean that when we have a malformed header line the following will happen:

If ignore_header_errors is False:
1. a logger.warning message is emitted here
2. an exception occurs here at line 772
3. In `read_header_section()`, line 559 the exception is caught an re-raised as a LASHeaderError
4. parsing stops

If ignore_header_errors is True:
1. a logger.warning message is emitted here
2. a logger.warning message is again emitted at line 557 in `read_header_section()`
3. parsing continues

If I'm correct, I think we should remove the logger.warning call here and accept that because an exception is raised there is no need for a logger.warning call. What do you think?

Reply:
Yes, That is right.  Removing the line 770 logger.warning in the next commit.  

One side note for ignore_headers_errors=False:
The exception occurring at line 772, first reports an `AttributeError: 'NoneType.. During Handling another exception occurred: ...` and then `LASHeaderError..`. So 2 exceptions will be displayed and the root cause is at the first one.  I think this is okay, but let me know if it should be changed.

Reply:
Good point! It would be better to have just the one exception raised from read_header_line(), I didn't really think too closely about it. That would be a good idea, if you're willing to rearrange it, but it's not necessary, I don't mind either way.

Reply:
I haven't found the way yet to arrange things to only raise the LASHeaderError and have it include the relevant AttributeError information.  So if you don't mind, I'll work it out in a future pull-request.

Reply:
That's totally fine! Thanks so much for all this!

Reply:
@kinverarity1,

I spent time today to look at ways of simplifying or combining the two exceptions `LASHeaderError` and `AttributeError` in  parse_header_section().  The outcome is that so far having both display is better than other solutions I found.  So I think we should leave them be unless some future information provides a better way.

Details:
- I could make one display and not the other but in those cases some useful debugging information was lost.  If only AttributeError  is displayed then the real line number of the error is in the stack trace but the line of las meta data being acted on is dropped.  If only LASHeaderError is displayed then the real line number of the error is dropped.

- There are way to programmatically combine the two exceptions but that added specific and unusual code to the exception handling and cluttered the regular flow of Lasio code with exception specific code.  So that doesn't seem like a good approach.

Given these finding, I won't make a PR for changing these two exceptions at this time. 

Let me know if you have additional thoughts on it.

Thanks!,
DC